Live reloading when switching themes

docrjp opened the pull request:
This resolves enhancement #7335 ( fixes #7335 )

Theme preferences can be changed without reloading. Supporting the enhancement is a refactoring of Theme which lays the groundwork for further enhancements #7322 and #7323.

This is a draft PR, for visibility and Continuous Integration purposes. Some code changes are outstanding:

- [X] Live reload when changing between light/dark/custom
- [x] Review live reloading for web engine (Preview Viewer)
- [x] If all live reloading is fully effective, remove reload warning
- [x] Rework ThemeTest now that the Theme interface has changed for web engine stylesheets
- [ ] Review JavaDoc and comments to explain solution
- [ ] Summarise solution here to assist review

Followed by standard PR tasks:

- [ ] Change in CHANGELOG.md described (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

tobiasdiez commented on src/main/java/org/jabref/gui/theme/StyleSheetFile.java:
```diff
@@ -0,0 +1,144 @@
+
+    @Override
+    void reload() {
+        embedDataUrl(url, dataUrl::set);
```
I would slightly refactor this to `getDataUrl(url).ifPresent(dataUrl::set)`.

tobiasdiez commented on src/main/java/org/jabref/gui/theme/ThemeImpl.java:
```diff
@@ -0,0 +1,219 @@
+
+        private final ThemePreference themePreference;
+
+        private final StyleSheet additionalStylesheet;
```
Maybe make this an `Optional<StyleSheet>`? Than you would not need to introduce the "EmptyStylesheet" class.

docrjp replied:
The idea here is to have an invariant: "there is always an additional stylesheet".

This is so that it is much simpler to transition between states where an additional stylesheet is available, and states where it is not (either because it is missing, or because the theme does not require one).

tobiasdiez replied:
ok, but "there is always an additional stylesheet" doesn't seem to be valid from a business logic (the light theme doesn't have additional css) as well as from the current implementation (EmptyStylesheet is not a a stylesheet, it's something that pretends to be one).

On the other hand, the current solution works and is fine codewise so take my comment as a suggestion that doesn't necessarily need to be implemented.

docrjp replied:
Ah, perhaps it wasn't clear! EmptyStylesheet loaded a real resource called Empty.css so it wasn't pretending!

However, given it is confusing, I've updated to remove StyleSheetEmpty and rename Empty.css to Light.css, which has a comment in it explaining that it doesn't have to add anything over Bas.css

This is meaningful, because you _could_ add something specific to Light.css. A practical example would be theming for Preview. That can't be done in Base.css, because WebEngine (which preview uses) can only take one style sheet and that is the additional one. So Light.css would be the place for that.

So hopefully that clears things up while keeping the useful invariant.

tobiasdiez commented on src/main/java/org/jabref/gui/theme/ThemeImpl.java:
```diff
@@ -0,0 +1,219 @@
+        if (updateRunner == null) {
+            throw new IllegalArgumentException("Update runner required");
+        }
+        this.preferencesService = preferencesService;
```
Use `this.preferencesService = Objects.requireNonNull(preferenceService)` instead of the manual null check above.

tobiasdiez reviewed (commented):
Thanks a lot, this already looks extremely nice. The new `Stylesheet` classes are a good idea, they make things a lot clearer.

I'm not sure if it works, but it would be nice to profit even more from the introduction of these classes and further simplify the Theme / ThemeImpl / ThemePreference classes. The following would be my idea:
- Add `Optional<Stylesheet> getAdditonialStylesheet` in `AppearencePreference`. The preference class uses `Stylesheet.create(...)` to create one based on the string stored in the preferences.
- The Theme class uses this method to determine the additional stylesheets to install.
- You can now remove ThemeState, ThemeImpl and ThemePreferences.

What do you think?

Siedlerchr commented on src/main/java/org/jabref/preferences/AppearancePreferences.java:
```diff
@@ -24,4 +26,23 @@ public int getMainFontSize() {
+            return false;
+        }
+        AppearancePreferences that = (AppearancePreferences) o;
+        return shouldOverrideDefaultFontSize == that.shouldOverrideDefaultFontSize &&
```
Tipp: You can use Object.equals(a,b)

docrjp replied:
Done

docrjp commented:
> Thanks a lot, this already looks extremely nice. The new `Stylesheet` classes are a good idea, they make things a lot clearer.
> 
> I'm not sure if it works, but it would be nice to profit even more from the introduction of these classes and further simplify the Theme / ThemeImpl / ThemePreference classes. The following would be my idea:
> 
> * Add `Optional<Stylesheet> getAdditonialStylesheet` in `AppearencePreference`. The preference class uses `Stylesheet.create(...)` to create one based on the string stored in the preferences.
> * The Theme class uses this method to determine the additional stylesheets to install.
> * You can now remove ThemeState, ThemeImpl and ThemePreferences.
> 
> What do you think?

I have gone somewhat in this direction, just pushed a change that cleans up the AppearancePreferences and eliminates ThemeState, by having the Stylesheet created by the ThemePreference. Theme no longer has a circular relationship with the PreferenceService. Instead it just needs AppearancePreferences and this is the state object.

It might be a bit cleaner still if I can move AppreancePreferences into the theme package, at the moment all the preference objects sit together but we really want it to live with ThemePreference so it can have some package-private methods to help ThemeImpl.

Not sure I could go further with removing the added classes without compromising the Single Responsibility Principle which are the rationale behind them:

* Theme (ThemeImpl) is responsible for live updating changes to appearance. It is a singleton so that it can transition between old and new preferences
* Stylesheet and its implementations abstract the different places that a stylesheet can be stored, and the complexities of WebEngine and Scene having different URL handling
* AppearancePreferences hold all the chosen preferences at a point in time that can be transitioned by Theme
* ThemePreference is the specific appearance preference that determines the chosen additional stylesheet. It is responsible for determining the correct additional Stylesheet for a chosen theme. This is the class that will be refactored as part of #7322

If we for example combined Stylesheet and ThemePreference into a single class, then Stylesheet would no longer be useable for the base stylesheet, it would be coupled to the additional stylesheet. The general stylesheet handling doesn't need to be impacted when we get to #7322, which really only concerns how ThemePreference  interacts with the appearance tab and settings storage. Stylesheet shouldn't have to care about such things, especially as it is also used by the base CSS.

tobiasdiez commented:
That looks really good. Thanks! Based on your description of the responsibilities of the classes, one might consider to rename `Theme` to ̀ ThemeManager` (and `ThemePreferences` to `Theme`). Also maybe it's possible to merge ThemePreference with AppearencePreference since the former is never used as a standalone anyway.

calixtus commented:
> It might be a bit cleaner still if I can move AppreancePreferences into the theme package, at the moment all the preference objects sit together but we really want it to live with ThemePreference so it can have some package-private methods to help ThemeImpl.

Great work so far, however, please keep in mind that the `XxxPreferences` objects should be plain stupid (immutable) data holder objects, that are created by a getter method by a `PreferencesService`. One of our goals is to clean up the whole codebase of JabRef, which has grown in almost 20 years, modularize it and implement design patterns. I know, this is still a very long road to walk...

calixtus commented:
Hi @docrjp , do you have any updates on this PR? We are so impressed of your work you already contributed in such a short period of time in this and the other PR.
Tonight is devcall, and we normally walk through the PRs to check for their status.

docrjp commented:
> Hi @docrjp , do you have any updates on this PR? We are so impressed of your work you already contributed in such a short period of time in this and the other PR.
> Tonight is devcall, and we normally walk through the PRs to check for their status.

Hi! So, I'm somewhat undecided about how to finish up the PR.

Eliminating ThemeState, and moving the Stylesheet static factory to the XxxPreferences, conflicts to some extent with the principle that the preferences are just data holder objects. Since now, the implementation of Stylesheet will depend upon the data. On the other hand, maybe that's taking the principle too literally. It's not too much different from java NIO Path. If we used Path in a preference class, then we're using static factory methods i.e. Path.of() and these static factories have a lot of file system logic behind them, with Path itself being an interface.

I think it depends on the goal behind the preference classes being data holders.

If it is so that we can deserialize by reflection with a library (e.g. Jackson), where it helps to not have complex factory or constructor logic, then I think it safest to move this logic back out of ThemePreference and reintroduce ThemeState as per the initial PR commit. Doing so eliminates the static factory from the ThemePreference and will keep that class really plain and simple.

On the other hand, if the purpose of keeping the preferences as data holder objects is to encourage loose coupling -- i.e. avoid the preferences needing application services injecting into them -- then the Stylesheet.create() static factory itself is harmless. It doesn't need to depend upon any other services, less so even than Path.of() which depends on file system providers.

However, what remains not so great is that Stylesheet is in the theme package but the factory is used from the preference package -- and the preference is in turn used by the ThemeManager in the .theme package. So that's a cyclic dependency of packages.

My instinct is to keep ThemePreference, and keep Stylesheet.create() as being called by ThemePreference, and then move ThemePreference into its own sub-package of the preferences package along with Stylesheet. But I don't want to do that if it messes up other plans with preferences.

calixtus commented:
Thanks for your very quick response!
The big plans were to uncouple the current implementation of JabRefPreferences from the other code and to clean up the project structure, so we reach someday a clean distinction between model logic and gui (and in the far far future in a galaxy far far away maybe someday provide a webbased interface for jabref).

My idea was to collect the xxxPreferences classes in the model package unbound to any concrete implementation either of the preferences or the gui. On the other hand I think that the theme preferences definitely are part of the gui implementation.
I think we'll discuss this in the devcall tonight.

tobiasdiez commented:
> the purpose of keeping the preferences as data holder objects is to encourage loose coupling -- i.e. avoid the preferences needing application services injecting into them

That's their main goal as of now. Personally, I don't mind if a bit of logic is needed to construct a preference object.


> However, what remains not so great is that Stylesheet is in the theme package but the factory is used from the preference package -- and the preference is in turn used by the ThemeManager in the .theme package. So that's a cyclic dependency of packages.

I don't think this is a big deal. For unit tests, you can mock the preferences and still create a ThemeManager. I think, this is the most important thing that one desires from the architecture.

docrjp commented:
Did anything come out of the devcall that is worth taking into consideration?

If not, then based on discussion above, I think we can just stick with what we've got for now, and I finish up the outstanding work of fixing the unit tests. I'll also update the tracked master to make sure that the PR is clean.

With the related follow up issues #7322 and #7323. there will be some more natural opportunities to refactor and discuss :)

calixtus commented:
Sorry, I was a bit buried in work today.

We had a brief discussion yesterday, but no quick response, so it's more or less up to me, since I'm currently refactoring the whole preferences complex.

I believe there was a misunderstanding on my side: I thought that ThemePreference is a preferences class just like MainTablePreferences, but it rather seems to be the representation of the actual theme. So please rename it to Theme, like @tobiasdiez mentioned earlier. I'd like to ask you to provide the ThemeManager in JabRefGui or in Globals, but not to construct it in JabRefPreferences, since there we only want to return values from the low level preferences repository. That way we can avoid bloating the JabRefPrefences too much with logic and gui stuff, that does not belong there. That's because my future target is to abstract the getXXXPrefrences and storeXXXPreferences methods in JabRefPreferences to be able to provide multiple implementations of a PreferencesService. So JabRefPreferences should only implement an adapter for the java prefs.

Please keep the naming convention (getAppearancePreferences instead of getCurrentAppearancePreferences). Also please use these methods only to store preferences or return from the java prefs or from a cache variable.

I know, JabRefPreferences is still a bit a messy place, but I'm working on it. It has a code history of almost 20 years now.

Thank you for your patience and all your impressive work!

As soon as you are ready, we are doing the regular double review and can then merge.

tobiasdiez commented:
@docrjp could you please implement the last remaining changes so that we can get this merged? Thanks!

docrjp commented:
Sorry! My spare time has been unusually lacking this last month but I hope
to get to it soon!

All the best
Rob


On Wed, 10 Mar 2021, 09:59 Tobias Diez, ***@***.***> wrote:

> @docrjp <https://github.com/docrjp> could you please implement the last
> remaining changes so that we can get this merged? Thanks!
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/JabRef/jabref/pull/7336#issuecomment-795179410>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACIZMN3X3VZ2V33CKJORJPTTC47G3ANCNFSM4V7P6JVQ>
> .
>

calixtus commented:
Hi, I took the liberty to dive into this PR a bit last week and made some commits regarding the remaining issues. If you don't mind I'll push them on your branch?

docrjp commented:
Sure!

On Wed, 10 Mar 2021, 11:38 Carl Christian Snethlage, <
***@***.***> wrote:

> Hi, I took the liberty to dive into this PR a bit last week and made some
> commits regarding the remaining issues. If you don't mind I'll push them on
> your branch?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/JabRef/jabref/pull/7336#issuecomment-795292049>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACIZMN4XW5KIVITJPU55JWLTC5K3TANCNFSM4V7P6JVQ>
> .
>

calixtus commented:
But I encountered something I did not really understand: Why should a style remain in the memory, if it has deleted from disk?

calixtus commented:
Still some problems with the ThemeManagerTest. Need to take a deeper look into this. Keeping stylesheets in memory seems not working properly.

docrjp commented:
Yes the tests not working is the main outstanding work and to be expected;
the assertions need updating given some of the new invariants

On Wed, 10 Mar 2021, 21:47 Carl Christian Snethlage, <
***@***.***> wrote:

> Still some problems with the ThemeManagerTest. Need to take a deeper look
> into this. Keeping stylesheets in memory seems not working properly.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/JabRef/jabref/pull/7336#issuecomment-796176959>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACIZMN47JT6KZF5AICLUJ2TTC7SIXANCNFSM4V7P6JVQ>
> .
>

docrjp commented:
Why was Light.css removed? This invariant that all themes have an
additional css is necessary for the live reloading to work which is the
purpose merge request.

On Thu, 11 Mar 2021, 00:23 Rob Platt, ***@***.***> wrote:

> Yes the tests not working is the main outstanding work and to be expected;
> the assertions need updating given some of the new invariants
>
> On Wed, 10 Mar 2021, 21:47 Carl Christian Snethlage, <
> ***@***.***> wrote:
>
>> Still some problems with the ThemeManagerTest. Need to take a deeper look
>> into this. Keeping stylesheets in memory seems not working properly.
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/JabRef/jabref/pull/7336#issuecomment-796176959>, or
>> unsubscribe
>> <https://github.com/notifications/unsubscribe-auth/ACIZMN47JT6KZF5AICLUJ2TTC7SIXANCNFSM4V7P6JVQ>
>> .
>>
>

docrjp commented:
> But I encountered something I did not really understand: Why should a style remain in the memory, if it has deleted from disk?

One less way to crash. I can't guarantee a file's existence between checking for its existence and passing the URL to web engine because race conditions.

The real question is why the web engine works differently to the rest of OpenJFX with respect to CSS caching, but that would be for OpenJFX to explain. We're stuck working with this specialness just for previews.

Which leads to the other reason: consistency. Rest of application stays themed if you delete the CSS, why not the web engine based preview?

Note that this solution isn't new to this MR, it is from previous accepted MR  https://github.com/JabRef/jabref/pull/7292

docrjp commented:
Also, the language entries for requiring a restart have come back.

These were removed as unused entries fail a quality gate and the point of this MR is to no longer need a restart when changing theme.

calixtus commented:
Hi @docrjp 
These are just a proposal open for discussion and certainly not done. I can comment a bit about my ideas tonight.

calixtus commented:
> Also, the language entries for requiring a restart have come back.

Maybe a merge error when merging upstream/master into your branch. I will take a look tonight.

docrjp commented:
Awesome.

Just to explain where I was up to so it is clear why the tests are failing.

I had fully tested manually the changes and it all works, can change themes without restart.

docrjp commented:
Whoops clumsy fingers

docrjp commented:
So to be clear the outstanding work was:

* Refactor ThemePreference to Theme
* Move ThemeManager to Globals (do note this class though renamed by me was already in JabRefPreferences so this refactor is added scope)
* Update outdated tests
* Remove getCurrentAppearancePreferences() altogether

I think you're banging your head against getCurrentAppearancePreferences() and trying to get tests to pass with it, but it's a dead method altogether.

calixtus commented:
Glad, my ideas were a bit helpful. 😃

docrjp commented:
> Glad, my ideas were a bit helpful. 😃

Always the end result is better for more minds :)

Siedlerchr commented:
I just upgraded to javafx16, this means it's no long necessary to create base64 data uris as @tobiasdiez fixed the bug https://bugs.openjdk.java.net/browse/JDK-8240969 for the WebEngine setUserStyleSheet

docrjp commented:
@calixtus how about this plan:

If you want to take a look as you fancy to get it where you want tonight and tomorrow night, I'll then take it forward this weekend.

That way we won't conflict on each other and hopefully we'll end the week being able to take this out if draft :-)

calixtus commented:
@docrjp Thanks, yes, plan sounds good. Just made one commit.
I removed the light.css file, because it felt 'unnatural' to have a file, that just exists for no reason but to avoid an additional if clause.
About the scope of the ThemeManager: I know, putting the creation of the ThemeManager into Globals doesn't really make the scope of the ThemeManager smaller, but on the other hand, we already have several Managers created in Globals and we are looking for a general smooth solution to replace Globals. I introduced DI for the ThemeManager, to make it a bit more pleasant. 😄 

Remaining bugs:
- ThemeManagerTests fail. I did not have time to look into this. I was vaccinated yesterday and the side effects are a little bit unpleasant.
- Switching to dark theme and back to light theme throws an out of bound exception. Somewhere two instances of the Dark.css stylesheet are stored additionally to the base.css in the stylesheet list. Again, no time to investigate.

docrjp commented:
The out of bounds is because without the Light.css, the observable list
size for CSS files changes and you'll need to do a lot more complicated
stuff with the observable list.

I think the opposite anyway: it is natural that every theme has its CSS
file. It's a matter of perspective and I discussed this already with
Tobias, I think we need Light.css back.

When I say scope, I mean the scope and size of the merge request, it is
getting really unwieldy: ThemeManager was in JabRefPreferences because it
was renamed from Theme which was already mangled with it.

I guess I am trying to say, and don't take this the wrong way, but the
merge request keeps becoming more than it needs to be. I say we on we moved
to Globals now but let's get to a conclusion, I had split up all the
recommendations into multiple issues to avoid precisely this. So let's
restore Light.css and I can focus on updating the tests then progress in
future updates. I hope that doesn't come across incorrectly!

On Sat, 13 Mar 2021, 11:20 Carl Christian Snethlage, <
***@***.***> wrote:

> @docrjp <https://github.com/docrjp> Thanks, yes, plan sounds good. Just
> made one commit.
> I removed the light.css file, because it felt 'unnatural' to have a file,
> that just exists for no reason but to avoid an additional if clause.
> About the scope of the ThemeManager: I know, putting the creation of the
> ThemeManager into Globals doesn't really make the scope of the ThemeManager
> smaller, but on the other hand, we already have several Managers created in
> Globals and we are looking for a general smooth solution to replace
> Globals. I introduced DI for the ThemeManager, to make it a bit more
> pleasant. 😄
>
> Remaining bugs:
>
>    - ThemeManagerTests fail. I did not have time to look into this. I was
>    vaccinated yesterday and the side effects are a little bit unpleasant.
>    - Switching to dark theme and back to light theme throws an out of
>    bound exception. Somewhere two instances of the Dark.css stylesheet are
>    stored additionally to the base.css in the stylesheet list. Again, no time
>    to investigate.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/JabRef/jabref/pull/7336#issuecomment-798183516>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACIZMN3C22Q64TZW3LA7KGTTDNC6VANCNFSM4V7P6JVQ>
> .
>

calixtus commented:
Ok, restore Light.css, but please keep the instancing of the ThemeManager out of the Preferences. You can probably force-push you branch and overwrite my commits.

docrjp commented:
> Ok, restore Light.css, but please keep the instancing of the ThemeManager out of the Preferences. You can probably force-push your branch and overwrite my commits.

Oh for sure you made many other nice changes too 🙂 so I won't force push over them. I will just put Light.css back.

I probably sounder stronger than I intended, I am commenting very fast because I should be decorating a bedroom 😜

calixtus commented:
Take your time, although I really would love to see this PR be merged into master before the next release. That makes theming JabRef so very much easier (and also development - you have no idea how frustrating it is recompile jabref over and over to see, if one tiny little annoying line is gone 🙈 ).

calixtus commented:
Hi @docrjp , does the bedroom already have some nice curtains and a bed? 😅

calixtus commented:
Hi @docrjp , i just wanted to ask, if you had the chance to make some progress here?
Right now we are still using java 14 for compiling and deploying jabref. I think as soon as we switch to java 16 we can simplify this PR a little bit thanks to a fix @tobiasdiez recently contributed to javafx.

Siedlerchr commented:
@calixtus javafx is independent of jdk and the new javafx version is already included since a while, see my earlier comment

calixtus commented:
> @calixtus javafx is independent of jdk and the new javafx version is already included since a while, see my earlier comment

You are right. Mixed both up in my thought process. 🙈 
Even better: we don't have to wait to switch to javafx 16. The future is now. 😄

koppor commented:
DevCall Decision: We really like this feature and would love it to get included it in the next release. We sadly did not see any activity the last weeks.

Proposal: @calixtus takes over for now. We would really love to see that @docrjp finds time to work on the remaining things (tests, ...) in a follow-up PR.

docrjp commented:
> DevCall Decision: We really like this feature and would love to include it in the next release. We sadly did not see any activity in the last weeks.
> 
> Proposal: @calixtus takes over for now. We would really love to see that @docrjp finds time to work on the remaining things (tests, ...) in a follow-up PR.

Go for it.

I've had a look but this would need serious time for me to get a handle on the many refactorings added in, to be comfortable giving the green light with not many automated tests. It is only my second MR. Desktop applications are not easily conducive to automated tests I know, but if I were to finish this MR it would be with those tests still in.

If I might offer some feedback: sometimes a small change is better even if reveals more opportunities for improvement and refactoring. This MR really got away from me in a way that I could only recover from with some very significant time investment.

koppor commented:
> I've had a look but this would need serious time for me to get a handle on the many refactorings added in, to be comfortable giving the green light with not many automated tests. 

Sometimes, a change requires serious architecture changes. 😇 

> It is only my second MR. 

Then we hope for more to come!

> Desktop applications are not easily conducive to automated tests I know, but if I were to finish this MR it would be with those tests still in.

It is especially difficult when files are coming in. Nevertheless, we should add test. In case you feel comfortable, you could start a MR adding some test cases at places where you would go for a change afterwards. We wrote some guidance on testing at https://devdocs.jabref.org/getting-into-the-code/code-howtos#test-cases.

koppor commented:
Discussion: There are rumors that JavaFX will provide a theming on its own. We need to get more information on that.

calixtus commented:
To think about: https://bugs.openjdk.java.net/browse/JDK-8267554

calixtus commented:
There still seems to be a concurrency bug producing the exception, and the ThemeManagerTest not completely working. Had to put work on this pr on hold because of more pressing matters, both private and jabref-related 
(observable prefs project)

Siedlerchr commented on src/main/java/org/jabref/gui/slr/ManageStudyDefinitionView.java:
```diff
@@ -68,9 +68,9 @@
+    @Inject private ThemeManager themeManager;
 
     private ManageStudyDefinitionViewModel viewModel;
     private final Study study;
```
checkstyle

Siedlerchr commented on src/main/java/org/jabref/gui/preferences/appearance/AppearanceTabViewModel.java:
```diff
@@ -82,54 +75,34 @@ public void setValues() {
             themeDarkProperty.setValue(false);
             themeCustomProperty.setValue(false);
-        } else if (currentTheme.getType() == Theme.Type.DARK) {
+        } else if (currentTheme.getType() == Theme.Type.EMBEDDED) {
```
Default and embedded are both embedded, I would name it Embedded_Light and embedded_dark

calixtus replied:
Light is not really a theme, it's rather the lack of existence of a theme.
EMBEDDED is to prepare more embedded themes (see the constructor of Theme), just by providing the filename as a static variable like "dark.css".

thedino-girl commented:
Is this feature ready now? I'm trying to make themes and would like to have this
